src/mavsdk/plugins/component_metadata/component_metadata_impl.cpp

```diff
@@ -0,0 +1,559 @@
+#include "component_metadata_impl.h"
+#include "callback_list.tpp"
+#include "fs_utils.h"
+#include <inflate_lzma.h>
+
+#include <utility>
+#include <filesystem>
+#include <fstream>
+#include <random>
+#include <regex>
+
+namespace mavsdk {
+
+namespace fs = std::filesystem;
+
+ComponentMetadataImpl::ComponentMetadataImpl(
+    System& system,
+    std::vector<uint8_t> components_to_request,
+    std::optional<std::string> translation_locale) :
+    PluginImplBase(system),
+    _components_to_request(std::move(components_to_request)),
+    _translation_locale(translation_locale)
+{
+    _system_impl->register_plugin(this);
+    init_object();
+}
+
+ComponentMetadataImpl::ComponentMetadataImpl(
+    std::shared_ptr<System> system,
+    std::vector<uint8_t> components_to_request,
+    std::optional<std::string> translation_locale) :
+    PluginImplBase(std::move(system)),
+    _components_to_request(std::move(components_to_request)),
+    _translation_locale(translation_locale)
+{
+    _system_impl->register_plugin(this);
+    init_object();
+}
+
+void ComponentMetadataImpl::init_object()
+{
+    const auto cache_dir_option = get_cache_directory();
+    if (cache_dir_option) {
+        _file_cache.emplace(cache_dir_option.value() / "component_metadata", 50);
+    } else {
+        LogErr() << "Failed to get cache directory";
+    }
+
+    const auto tmp_option = create_tmp_directory("mavsdk-component-metadata");
+    if (tmp_option) {
+        _tmp_download_path = tmp_option.value();
+    } else {
+        _tmp_download_path = "./mavsdk-component-metadata";
+        std::error_code err;
+        std::filesystem::create_directory(_tmp_download_path, err);
+    }
+}
+
+ComponentMetadataImpl::~ComponentMetadataImpl()
+{
+    _system_impl->unregister_plugin(this);
+    std::error_code ec;
+    std::filesystem::remove_all(_tmp_download_path, ec);
+    if (ec) {
+        LogErr() << "failed to remove directory: " << ec.message();
+    }
+}
+
+void ComponentMetadataImpl::init() {}
+
+void ComponentMetadataImpl::deinit() {}
+
+void ComponentMetadataImpl::enable()
+{
+    _mavlink_components.clear();
+    for (const auto& component : _components_to_request) {
+        _mavlink_components[component] = MavlinkComponent{};
+        _system_impl->request_message().request(
+            MAVLINK_MSG_ID_COMPONENT_METADATA, component, [this](auto&& result, auto&& message) {
+                receive_component_metadata(result, message);
+            });
+    }
+}
+
+void ComponentMetadataImpl::disable()
+{
+    // TODO: stop ongoing downloads?
```

Comment:
I would think we should do so, if possible.

Reply:
This is a bit more work I think. I might do it in a follow-up if you're ok.